### PR TITLE
Restrict the scope of cancel API

### PR DESCRIPTION
**Pull request by penghuo**
### Description
* For select query
  * if query result state is not success | fail | cancelled, the query can be cancelled. Todo, currently running query can not been cancelled https://github.com/opensearch-project/opensearch-spark/issues/243.

* For Flint index management query
  * `CREATE SI/CI/MV (auto_refresh=true)` can not be cancelled, error message is "Can not cancel index DML query"
  * `REFRESH SI/CI/MV` can be cancelled.
 
* OpenSearch 2.12 Upgrade Impact
  * if jobType is null AsyncQueryMetadata, BatchHandler handle cancel query as same as https://github.com/opensearch-project/sql/issues/2430
  * No impact for interactive query.

### Issues Resolved
https://github.com/opensearch-project/sql/issues/2430
 
### Check List
- [x] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/BatchQueryHandler.java [hunk 1]
```diff
@@ -57,9 +57,9 @@ protected JSONObject getResponseFromExecutor(AsyncQueryJobMetadata asyncQueryJob
 
   @Override
   public String cancelJob(AsyncQueryJobMetadata asyncQueryJobMetadata) {
```
Comment (vamsi-amazon) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/BatchQueryHandler.java:59]: This handles StreamingQueryHandler as well right.

**Review by YANG-DB — APPROVED** [review]

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/InteractiveQueryHandler.java [hunk 2]
```diff
@@ -62,7 +63,13 @@ protected JSONObject getResponseFromExecutor(AsyncQueryJobMetadata asyncQueryJob
...
   public String cancelJob(AsyncQueryJobMetadata asyncQueryJobMetadata) {
     String queryId = asyncQueryJobMetadata.getQueryId().getId();
-    getStatementByQueryId(asyncQueryJobMetadata.getSessionId(), queryId).cancel();
+    Statement statement = getStatementByQueryId(asyncQueryJobMetadata.getSessionId(), queryId);
+    if (NOT_CANCELLABLE_STATE.contains(statement.getStatementState())) {
+      throw new IllegalArgumentException(
+          String.format(
+              "can't cancel statement in %s state", statement.getStatementState().getState()));
+    }
+    statement.cancel();
```
Comment (noCharger) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/InteractiveQueryHandler.java:72]: it's already here 

https://github.com/opensearch-project/sql/blob/main/spark/src/main/java/org/opensearch/sql/spark/execution/statement/Statement.java#L67-L76

Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/InteractiveQueryHandler.java]: thanks.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/BatchQueryHandler.java (hunk 1 above)
Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/BatchQueryHandler.java]: refactor code, support cancel refresh statement in latest commit.

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java [hunk 3]
```diff
@@ -117,6 +123,17 @@ public String cancelJob(AsyncQueryJobMetadata asyncQueryJobMetadata) {
           new InteractiveQueryHandler(sessionManager, jobExecutionResponseReader, leaseManager);
     } else if (IndexDMLHandler.isIndexDMLQuery(asyncQueryJobMetadata.getJobId())) {
       queryHandler = createIndexDMLHandler(emrServerlessClient);
+    } else if (asyncQueryJobMetadata.getJobType() == JobType.BATCH) {
+      queryHandler =
+          new RefreshQueryHandler(
```
Comment (seankao-az) [inline spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:128]: JobType.BATCH always means it's a refresh job? When is the `BatchQueryHandler` on L139 used?

**Review by dai-chen — APPROVED** [review]
Thanks for the changes!

File: spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java (hunk 3 above)
Comment (penghuo) [reply spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java:128]: Yes, always refresh Job.
BatchQueryHandler it is used by UT only. we can refactor this part.

**Review by vamsi-amazon — APPROVED** [review]

**Pull request merged**